File: RemoveLoopFromLinkedList/RemoveLoopFromLinkedList.cpp
// RemoveLoopFromLinkedList.cpp : This file contains the 'main' function. Program execution begins and ends there.
//

#include <iostream>
#include <unordered_set>

class ListNode
{
public:
    int data;
    ListNode* next;
    ListNode(int data)
        : data(data)
        , next(nullptr)
    {}
};

ListNode* GetListFromInput(ListNode** ppHead, int count)
{
    if (count == 0)
        return nullptr;

    std::cout << "Enter " << count << " numbers of the list: ";
    ListNode* pTail = *ppHead;
    for (int i = 0; i < count; ++i)
    {
        int data = 0;
        std::cin >> data;
        if (*ppHead == nullptr)
        {
            *ppHead = new ListNode(data);
            pTail = *ppHead;
        }
        else
        {
            pTail->next = new ListNode(data);
            pTail = pTail->next;
        }
    }

    return pTail;
}

void PrintList(ListNode* pHead)
{
    if (pHead == nullptr)
        return;

    std::cout << pHead->data;
    pHead = pHead->next;

    while (pHead != nullptr)
    {
        std::cout << ", " << pHead->data;
        pHead = pHead->next;
    }
}

void FreeList(ListNode* pNode)
{
    if (pNode == nullptr)
        return;

    FreeList(pNode->next);
    delete pNode;
}

ListNode* GetNodeByIndex(ListNode* pHead, int index)
{
    if (index == -1)
        return nullptr;

    for (int i = 0; i < index; ++i)
    {
        pHead = pHead->next;
    }

    return pHead;
}

ListNode* EndNodeOfLoop(ListNode* list)
{
    std::unordered_set<ListNode*> nodes;
    ListNode* lastNode = nullptr;

    while (list != nullptr)
    {
        if (nodes.find(list) != nodes.cend())
            return lastNode;

        nodes.insert(list);
        lastNode = list;
        list = list->next;
    }

    return nullptr;
}

void RemoveLoop(ListNode* list)
{
    ListNode* pTail = EndNodeOfLoop(list);
    if (pTail != nullptr)
        pTail->next = nullptr;
}

int main()
{
    while (true)
    {
        int count = 0;
        std::cout << "Number of nodes in list: ";
        std::cin >> count;
        if (count < 0)
            break;

        ListNode* pHead = nullptr;
        ListNode* pTail = GetListFromInput(&pHead, count);

        std::cout << "The index of starting node of the loop: ";
        int index = -1;
        std::cin >> index;
        if (pHead != 0)
            pTail->next = GetNodeByIndex(pHead, index);

        RemoveLoop(pHead);

        std::cout << "The list after loop removed: ";
        PrintList(pHead);
        std::cout << std::endl;

        FreeList(pHead);
    }
}

File: Go/RemoveLoopFromLinkedList.go
package main

import "fmt"

type ListNode struct {
	data int
	next *ListNode
}

func GetArrayFromInput(count int, prompt string) []int {
	if count == 0 {
		return []int{}
	}

	fmt.Print(prompt)

	v := make([]int, count)
	for i := 0; i < count; i++ {
		fmt.Scan(&v[i])
	}

	return v
}

func GetListFromArray(v []int) (*ListNode, *ListNode) {
	if len(v) == 0 {
		return nil, nil
	}

	if len(v) == 1 {
		pNode := &ListNode{v[0], nil}
		return pNode, pNode
	}

	pHead := &ListNode{v[0], nil}
	pCur := pHead
	for i := 1; i < len(v); i++ {
		pCur.next = &ListNode{v[i], nil}
		pCur = pCur.next
	}

	return pHead, pCur
}

func (pNode *ListNode) Print() {
	fmt.Print(pNode.data)
	if pNode.next != nil {
		fmt.Print(", ")
	}
	if pNode.next != nil {
		pNode.next.Print()
	} else {
		fmt.Println("")
	}
}

func GetNodeByIndex(pHead *ListNode, index int) *ListNode {
	if index == -1 {
		return nil
	}

	for i := 0; i < index; i++ {
		pHead = pHead.next
	}

	return pHead
}

func GetStartingNodeOfLoop(list *ListNode) *ListNode {
	checked := make(map[*ListNode]struct{})
	for ; list != nil; list = list.next {
		_, exists := checked[list]
		if exists {
			return list
		}
		checked[list] = struct{}{}
	}

	return nil
}

func RemoveLoop(list *ListNode) {
	pNode := GetStartingNodeOfLoop(list)
	if pNode == nil {
		return
	}

	p := pNode
	for {
		if p.next == pNode {
			p.next = nil
			break
		}

		p = p.next
	}
}

func main() {
	for {
		fmt.Print("Number of elements: ")
		var count int
		fmt.Scan(&count)
		if count == 0 {
			break
		}

		vec := GetArrayFromInput(count, "Please enter the list: ")
		pHead, pTail := GetListFromArray(vec)

		fmt.Print("The node index the loop start with: ")
		var m int
		fmt.Scan(&m)

		pTail.next = GetNodeByIndex(pHead, m)

		RemoveLoop(pHead)

		fmt.Print("After loop removed, the list is ")
		pHead.Print()
	}
}
